oci/layout/oci_delete.go

```diff
@@ -0,0 +1,245 @@
+			if err != nil {
+				return err
+			}
+			index, err := parseIndex(blobPath)
+			if err != nil {
+				return err
+			}
+			err = ref.getBlobsUsedInIndex(destination, index, sharedBlobsDir)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unsupported mediaType in index: %q", descriptor.MediaType)
+		}
+	}
+
+	return nil
+}
+
+func (ref ociReference) getBlobsUsedInManifest(manifest *imgspecv1.Manifest) map[digest.Digest]int {
+	blobsUsedInManifest := make(map[digest.Digest]int, 0)
+
+	blobsUsedInManifest[manifest.Config.Digest]++
+	for _, layer := range manifest.Layers {
+		blobsUsedInManifest[layer.Digest]++
+	}
+
+	return blobsUsedInManifest
+}
+
+// This takes in a map of the digest and their usage count in the manifest to be deleted
+// It will compare it to the digest usage in the root index, and return a set of the blobs that can be safely deleted
+func (ref ociReference) getBlobsToDelete(blobsUsedByDescriptorToDelete map[digest.Digest]int, sharedBlobsDir string) (*set.Set[digest.Digest], error) {
+	rootIndex, err := ref.getIndex()
+	if err != nil {
+		return nil, err
+	}
+	blobsUsedInRootIndex := make(map[digest.Digest]int)
+	err = ref.getBlobsUsedInIndex(blobsUsedInRootIndex, rootIndex, sharedBlobsDir)
+	if err != nil {
+		return nil, err
+	}
+
+	blobsToDelete := set.New[digest.Digest]()
+
+	for digest, count := range blobsUsedInRootIndex {
+		if count-blobsUsedByDescriptorToDelete[digest] == 0 {
+			blobsToDelete.Add(digest)
+		}
+	}
+
+	return blobsToDelete, nil
+}
+
+func (ref ociReference) deleteBlobs(blobsToDelete *set.Set[digest.Digest]) error {
+	for _, digest := range blobsToDelete.Values() {
+		blobPath, err := ref.blobPath(digest, "") //Only delete in the local directory, not in the shared blobs path
+		if err != nil {
+			return err
+		}
+		err = deleteBlob(blobPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func deleteBlob(blobPath string) error {
+	logrus.Debug(fmt.Sprintf("Deleting blob at %q", blobPath))
+
+	err := os.Remove(blobPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logrus.Info(fmt.Sprintf("Blob at %q not found; it was either previously deleted or is in the shared blobs directory", blobPath))
+			return nil
+		} else {
+			return err
+		}
+	} else {
+		return nil
+	}
+}
+
+func (ref ociReference) deleteReferenceFromIndex(referenceIndex int) error {
+	index, err := ref.getIndex()
+	if err != nil {
+		return err
+	}
+
+	if ref.image == "" && len(index.Manifests) == 1 {
+		index.Manifests = make([]imgspecv1.Descriptor, 0)
+		return saveJSON(ref.indexPath(), index)
+	}
+
+	newDescriptors := make([]imgspecv1.Descriptor, 0, len(index.Manifests)-1)
+	for i, descriptor := range index.Manifests {
+		// Double check: same image reference name and same position in the index
+		if descriptor.Annotations[imgspecv1.AnnotationRefName] != ref.image || referenceIndex != i {
+			newDescriptors = append(newDescriptors, descriptor)
+		}
+	}
+	index.Manifests = newDescriptors
```

Comment:
I’d prefer just a `index.Manifests = slices.Delete(index.Manifests, referenceIndex, referenceIndex+1)` here. No special case for a single entry, no duplicate logic to maintain.

Alternatively, if you really prefer the double check here, please link the two functions with comments:
```go
// Update also ociReference.getManifestDescriptor
```
before starting to edit `index.Manifests` here, and a mirror comment in `getManifestDescriptor` pointing to this function.

Reply:
Ok, I see what you mean. I guess it depends on how comfortable we are on relying on the index alone. I was thinking that the code could check what's at that index before the final deletion, just to make sure, but judging from what you say, this is overkill.

Reply:
I don’t think there’s a way to be _perfect_ here, notably against a concurrent modification[1]. We could plausibly read the top-level index _once_ into memory, do a lookup on the in-memory data, delete in the in-memory data, and write the index back, to avoid deleting an unexpected entry if a concurrent process reshuffles the index in the meantime; but that would still not make anything _reliable_ because the concurrent process could also add a new reference to a blob we intend to delete.

The transport’s writes already only work correctly with an externally-implemented guarantee that there is at most one writer at a time, and this is no different — so I think we might just as well use a simple implementation.


[1] Cross-platform file locking, or locking on UNIX from inside a library, is a mess, so the transport doesn’t do any.